File: dotnet/Gherkin/ParserException.cs
using Gherkin.Ast;

namespace Gherkin;

public abstract class ParserException(string message, Location location = null) : Exception(GetMessage(message, location))
{
    public Location Location { get; } = location;

    private static string GetMessage(string message, Location location)
    {
        if (location == null)
            return message;

        return string.Format("({0}:{1}): {2}", location.Line, location.Column, message);
    }

}

public class AstBuilderException(string message, Location location) : ParserException(message, location)
{
}

public class NoSuchLanguageException : ParserException
{
    public NoSuchLanguageException(string language, Location location = null) :
        base("Language not supported: " + language, location)
    {
        if (language == null) throw new ArgumentNullException("language");
    }

}

public class InvalidTagException(string message, Location location = null) : ParserException(message, location)
{
}

public abstract class TokenParserException : ParserException
{
    protected TokenParserException(string message, Token receivedToken)
        : base(message, GetLocation(receivedToken))
    {
        if (receivedToken == null) throw new ArgumentNullException("receivedToken");
    }

    private static Location GetLocation(Token receivedToken)
    {
        return receivedToken.IsEOF || receivedToken.Location.Column > 1
            ? receivedToken.Location
            : new Location(receivedToken.Location.Line, receivedToken.Line.Indent + 1);
    }

}

public class UnexpectedTokenException : TokenParserException
{
    public string StateComment { get; set; }

    public Token ReceivedToken { get; set; }
    public string[] ExpectedTokenTypes { get; set; }

    public UnexpectedTokenException(Token receivedToken, string[] expectedTokenTypes, string stateComment)
        : base(GetMessage(receivedToken, expectedTokenTypes), receivedToken)
    {
        if (receivedToken == null) throw new ArgumentNullException("receivedToken");
        if (expectedTokenTypes == null) throw new ArgumentNullException("expectedTokenTypes");

        ReceivedToken = receivedToken;
        ExpectedTokenTypes = expectedTokenTypes;
        StateComment = stateComment;
    }

    private static string GetMessage(Token receivedToken, string[] expectedTokenTypes)
    {
        return string.Format("expected: {0}, got '{1}'",
            string.Join(", ", expectedTokenTypes),
            receivedToken.GetTokenValue().Trim());
    }

}

public class UnexpectedEOFException : TokenParserException
{
    public string StateComment { get; }
    public string[] ExpectedTokenTypes { get; }
    public UnexpectedEOFException(Token receivedToken, string[] expectedTokenTypes, string stateComment)
        : base(GetMessage(expectedTokenTypes), receivedToken)
    {
        if (receivedToken == null) throw new ArgumentNullException("receivedToken");
        if (expectedTokenTypes == null) throw new ArgumentNullException("expectedTokenTypes");

        ExpectedTokenTypes = expectedTokenTypes;
        StateComment = stateComment;
    }

    private static string GetMessage(string[] expectedTokenTypes)
    {
        return string.Format("unexpected end of file, expected: {0}",
            string.Join(", ", expectedTokenTypes));
    }
}

public class CompositeParserException : ParserException
{
    public IEnumerable<ParserException> Errors { get; }

    public CompositeParserException(ParserException[] errors)
        : base(GetMessage(errors))
    {
        if (errors == null) throw new ArgumentNullException("errors");

        Errors = errors;
    }

    private static string GetMessage(ParserException[] errors)
    {
        return "Parser errors:" + Environment.NewLine + string.Join(Environment.NewLine, errors.Select(e => e.Message));
    }
}

File: java/src/main/java/io/cucumber/gherkin/ParserException.java
package io.cucumber.gherkin;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import io.cucumber.messages.types.Location;

import static io.cucumber.gherkin.Locations.COLUMN_OFFSET;
import static io.cucumber.gherkin.Locations.atColumn;

class ParserException extends RuntimeException {
    final Location location;

    protected ParserException(String message) {
        super(message);
        location = null;
    }

    protected ParserException(String message, Location location) {
        super(createMessage(message, location));
        this.location = location;
    }

    private static String createMessage(String message, Location location) {
        if (location == null) {
            return String.format("(-1,0): %s", message);
        }
        Integer line = location.getLine();
        Integer column = location.getColumn().orElse(0);
        return String.format("(%s:%s): %s", line, column, message);
    }

    static class AstBuilderException extends ParserException {
        AstBuilderException(String message, Location location) {
            super(message, location);
        }
    }

    static class NoSuchLanguageException extends ParserException {
        NoSuchLanguageException(String language, Location location) {
            super("Language not supported: " + language, location);
        }
    }

    static class UnexpectedTokenException extends ParserException {
        String stateComment;

        final Token receivedToken;
        final List<String> expectedTokenTypes;

        UnexpectedTokenException(Token receivedToken, List<String> expectedTokenTypes, String stateComment) {
            super(getMessage(receivedToken, expectedTokenTypes), getLocation(receivedToken));
            this.receivedToken = receivedToken;
            this.expectedTokenTypes = expectedTokenTypes;
            this.stateComment = stateComment;
        }

        private static String getMessage(Token receivedToken, List<String> expectedTokenTypes) {
            return String.format("expected: %s, got '%s'",
                    String.join(", ", expectedTokenTypes),
                    receivedToken.getTokenValue()
            );
        }

        private static Location getLocation(Token receivedToken) {
            if (receivedToken.location.getColumn().isPresent()) {
                return receivedToken.location;
            }
            int column = COLUMN_OFFSET + receivedToken.line.getIndent();
            return atColumn(receivedToken.location, column);
        }
    }

    static class UnexpectedEOFException extends ParserException {
        final String stateComment;
        final List<String> expectedTokenTypes;

        UnexpectedEOFException(Token receivedToken, List<String> expectedTokenTypes, String stateComment) {
            super(getMessage(expectedTokenTypes), receivedToken.location);
            this.expectedTokenTypes = expectedTokenTypes;
            this.stateComment = stateComment;
        }

        private static String getMessage(List<String> expectedTokenTypes) {
            return String.format("unexpected end of file, expected: %s",
                    String.join(", ", expectedTokenTypes));
        }
    }

    static class CompositeParserException extends ParserException {
        final List<ParserException> errors;

        CompositeParserException(List<ParserException> errors) {
            super(getMessage(errors));
            this.errors = Collections.unmodifiableList(errors);
        }

        private static String getMessage(List<ParserException> errors) {
            if (errors == null) throw new NullPointerException("errors");
            return "Parser errors:\n" + errors.stream()
                    .map(Throwable::getMessage)
                    .collect(Collectors.joining("\n"));
        }
    }
}
